File: 0143-reorder-list/0143-reorder-list.js
var reorderList = function(head) {
    var lists = [], cur = head, prev;
    while(cur) {
        prev = cur;
        cur = cur.next;
        prev.next = null;
        lists.push(prev);
    };

    var left = 0, right = lists.length - 1, dummy = new ListNode();

    while(left < right) {
        lists[left].next = lists[right];
        dummy.next = lists[left];
        dummy = dummy.next.next; 
        left++;
        right--;
    }

    if(left == right) dummy.next = lists[left];
};

File: 0143-reorder-list/0143-reorder-list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return

        # Step 1: Find the middle
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        # Step 2: Reverse the second half
        prev, curr = None, slow.next
        while curr:
            next_temp = curr.next
            curr.next = prev
            prev = curr
            curr = next_temp
        slow.next = None  # Split the list into two halves

        # Step 3: Merge the two halves
        first, second = head, prev
        while second:
            temp1, temp2 = first.next, second.next
            first.next = second
            second.next = temp1
            first, second = temp1, temp2
